src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -115,24 +114,80 @@ void CoreWorkerDirectTaskSubmitter::OnWorkerIdle(const WorkerAddress &addr,
                                                  bool was_error) {
   absl::MutexLock lock(&mu_);
   if (queued_tasks_.empty() || was_error) {
-    RAY_CHECK_OK(lease_client_.ReturnWorker(addr.second));
+    auto lease_client = std::move(worker_to_lease_client_[addr]);
+    worker_to_lease_client_.erase(addr);
+    RAY_CHECK_OK(lease_client->ReturnWorker(addr.second));
   } else {
     auto &client = *client_cache_[addr];
     PushNormalTask(addr, client, queued_tasks_.front());
     queued_tasks_.pop_front();
   }
-  // We have a queue of tasks, try to request more workers.
-  if (!queued_tasks_.empty()) {
-    RequestNewWorkerIfNeeded(queued_tasks_.front());
-  }
+  RequestNewWorkerIfNeeded(queued_tasks_.front());
 }
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
-    const TaskSpecification &resource_spec) {
+    const TaskSpecification &resource_spec, const rpc::Address *raylet_address) {
   if (worker_request_pending_) {
     return;
   }
-  RAY_CHECK_OK(lease_client_.RequestWorkerLease(resource_spec));
+  if (queued_tasks_.empty()) {
+    // We don't have any tasks to run, so no need to request a worker.
+    return;
+  }
+
+  std::shared_ptr<WorkerLeaseInterface> lease_client;
+  if (raylet_address) {
+    // Connect to raylet.
+    ClientID raylet_id = ClientID::FromBinary(raylet_address->raylet_id());
+    auto it = remote_lease_clients_.find(raylet_id);
+    if (it == remote_lease_clients_.end()) {
+      RAY_LOG(DEBUG) << "Connecting to raylet " << raylet_id;
+      it =
+          remote_lease_clients_.emplace(raylet_id, lease_client_factory_(*raylet_address))
+              .first;
+    }
+    RAY_LOG(DEBUG) << "Sending " << resource_spec.TaskId() << " to raylet " << raylet_id;
+    lease_client = it->second;
+  } else {
+    lease_client = local_lease_client_;
+  }
+
+  // NOTE(swang): We must copy the resource spec here because the resource spec
+  // may get swapped out by the time the callback fires.
+  TaskSpecification resource_spec_copy(resource_spec.GetMessage());
```

Comment:
How about holding a shared_ptr of the TaskSpec? Copying the TaskSpec could be expensive because of protobuf serialization/deserialization cost. Previously I noticed non-trivial performance differences because of this when working on direct actor calls.

Reply:
Unfortunately, it's complicated to do in this case because the memory of the shared_ptr gets swapped out later on if the task is popped from the queue and sent to a worker before we hear back from the worker lease request. I figured it's better to copy here than to copy when we send the task (we should be sending more tasks than lease requests).

If we change the design so that we associate the lease request with this particular task, then we should be able to avoid the copy. I'll add this to the note.

Reply:
sounds good, would be nice to have a TODO.